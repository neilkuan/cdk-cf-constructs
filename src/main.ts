import * as fs from 'fs';
import * as path from 'path';
import {
  Stack,
  aws_lambda,
  aws_lambda_nodejs,
  aws_iam,
  custom_resources,
  aws_logs,
  CustomResource,
  Token,
  region_info,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface ICustomResourceGetEIPOptions {
  readonly companyIps?: string[];
  readonly regions?: string[];

  /**
   * Indicate whether always update the custom resource to get the new stack output
   * @default true
   */
  readonly alwaysUpdate?: boolean;
}
export class CustomResourceGetEIP extends Construct {
  outputs: CustomResource;
  constructor(scope: Construct, id: string, props?: ICustomResourceGetEIPOptions) {
    super(scope, id);
    const onEvent = new aws_lambda_nodejs.NodejsFunction(this, `${id}GetEIP`, {
      entry: fs.existsSync(path.join(__dirname, 'lambda/get-eip-handler.ts')) ? path.join(__dirname, 'lambda/get-eip-handler.ts') : path.join(__dirname, 'lambda/get-eip-handler.js'),
      handler: 'handler',
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      bundling: {
        forceDockerBundling: false,
        minify: true,
      },
    });

    const myProvider = new custom_resources.Provider(this, `${id}MyProvider`, {
      onEventHandler: onEvent,
      logRetention: aws_logs.RetentionDays.ONE_DAY,
    });

    onEvent.addToRolePolicy(new aws_iam.PolicyStatement({
      actions: ['ec2:DescribeAddresses'],
      resources: ['*'],
    }));

    if (props?.regions) {
      let right_regions: string[] =[];
      region_info.RegionInfo.regions.forEach(r=>{
        right_regions.push(r.name);
      });
      props.regions.forEach(r=>{
        if (!right_regions.includes(r)) {
          throw new Error(`Fault region code ${r}`);
        }
      });
    }

    this.outputs = new CustomResource(this, `${id}CustomGetEIP`, {
      resourceType: 'Custom::GetEIP',
      serviceToken: myProvider.serviceToken,
      properties: {
        STACK_NAME: Stack.of(this).stackName,
        COMPANY_IPS: Stack.of(this).toJsonString(props?.companyIps),
        REGIONS: Stack.of(this).toJsonString(props?.regions),
        randomString: props?.alwaysUpdate === false ? undefined : randomString(),
      },
    });

  }

  /**
   *
   * @returns Token.asList(this.outputs.getAtt('IP_LIST'));
   *
   */
  public ipList() {
    return Token.asList(this.outputs.getAtt('IP_LIST'));
  }
}

function randomString() {
  // Crazy
  return Math.random().toString(36).replace(/[^a-z0-9]+/g, '');
}