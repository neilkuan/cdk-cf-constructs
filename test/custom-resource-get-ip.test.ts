import { assertions, Stack, App } from 'aws-cdk-lib';
import { CustomResourceGetEIP } from '../src';


let stack: Stack;
beforeEach(() => {
  const app = new App();
  stack = new Stack(app, 'TestStack');
});

test('CustomResourceGetEIP testing', () => {

  const cr =new CustomResourceGetEIP(stack, 'TestCR', {
    regions: ['us-east-1'],
    companyIps: ['1.2.3.4'],
  });
  cr.getIps();


  expect(assertions.Template.fromStack(stack).toJSON()).toMatchSnapshot();
});


test('CustomResourceGetEIP testing not correct region', () => {
  expect(()=>{
    new CustomResourceGetEIP(stack, 'TestCR', {
      regions: ['us-east-99'],
      companyIps: ['1.2.3.4'],
    });
  }).toThrowError('Fault region code us-east-99');
});