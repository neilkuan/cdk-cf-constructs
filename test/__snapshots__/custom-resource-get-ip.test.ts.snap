// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomResourceGetEIP testing 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestCRTestCRCustomGetEIP8BE37B05": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "COMPANY_IPS": "[\\"1.2.3.4\\"]",
        "REGIONS": "[\\"us-east-1\\"]",
        "STACK_NAME": "TestStack",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestCRTestCRMyProviderframeworkonEvent70409DB7",
            "Arn",
          ],
        },
      },
      "Type": "Custom::GetEIP",
      "UpdateReplacePolicy": "Delete",
    },
    "TestCRTestCRGetEIP90A9240C": Object {
      "DependsOn": Array [
        "TestCRTestCRGetEIPServiceRoleDefaultPolicyBF26AF3E",
        "TestCRTestCRGetEIPServiceRole0483231C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5ae20a21685e9e7607c6c3f70c059c48e01c7decdb578251ef7763a6d6db4c20.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestCRTestCRGetEIPServiceRole0483231C",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestCRTestCRGetEIPServiceRole0483231C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestCRTestCRGetEIPServiceRoleDefaultPolicyBF26AF3E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ec2:DescribeAddresses",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestCRTestCRGetEIPServiceRoleDefaultPolicyBF26AF3E",
        "Roles": Array [
          Object {
            "Ref": "TestCRTestCRGetEIPServiceRole0483231C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestCRTestCRMyProviderframeworkonEvent70409DB7": Object {
      "DependsOn": Array [
        "TestCRTestCRMyProviderframeworkonEventServiceRoleDefaultPolicy527F64AB",
        "TestCRTestCRMyProviderframeworkonEventServiceRoleE9EA3243",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e7c95e1a12d903d913ebdd526c735f1ac58b2024882a5562110221fe685ebe85.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (TestStack/TestCR/TestCRMyProvider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "TestCRTestCRGetEIP90A9240C",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestCRTestCRMyProviderframeworkonEventServiceRoleE9EA3243",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestCRTestCRMyProviderframeworkonEventLogRetentionA1B56346": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "TestCRTestCRMyProviderframeworkonEvent70409DB7",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "TestCRTestCRMyProviderframeworkonEventServiceRoleDefaultPolicy527F64AB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestCRTestCRGetEIP90A9240C",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestCRTestCRGetEIP90A9240C",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestCRTestCRMyProviderframeworkonEventServiceRoleDefaultPolicy527F64AB",
        "Roles": Array [
          Object {
            "Ref": "TestCRTestCRMyProviderframeworkonEventServiceRoleE9EA3243",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestCRTestCRMyProviderframeworkonEventServiceRoleE9EA3243": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
